README.md
<div align="center">
  <h3 align="center">Breeze</h3>

  <p align="center">
    An awesome way to display your discord activity !
    <br />
    <a href="https://github.com/crizmo/breeze/blob/main/server-dcez/README.md"><strong>Explore more »</strong></a>
    <br />
    <a href="https://github.com/crizmo/breeze/tree/main/server-dcez">Server Side</a>
    ·
    <a href="https://github.com/crizmo/dcez-client">Client Side</a>
    ·
    <a href="https://breeze-stats.netlify.app/">Website</a>
  </p>
  <img src="https://api-breeze.herokuapp.com/api/compact/784141856426033233?banner=https://wallpaperaccess.com/full/6276627.jpg" />
</div>

<details>
  <summary>Features</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#issues">Issues</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>

## About The Project

<!-- [![Breeze][]]() -->

Breeze helps you display your discord activity on your website / anywhere you want using the api or the client.

Here's what breeze offers:
* client (website): display your discord activity on the website
* server (backend): display your discord activity as a api image which u can use on your website or anywhere you want

<p align="right">(<a href="#top">back to top</a>)</p>

## Usage

Breeze is very easy to use.
You can either use the api endpoint to display your discord activity or get your discord activity using the breeze client website.

You can display your discord activity as a card or a compact card.
The endpoint are the following:
<a href="https://api-breeze.herokuapp.com/api/card/:id">User card large</a> <br>
<a href="https://api-breeze.herokuapp.com/api/compact/:id">User card compact</a> <br>
Replace :id with your discord id + you can add the optional queries too.
The queries are:

  * banner: url of the banner image
  * hex: hex color for alternative banner color [ ?hex=D0D1D1 ]
  * about: about the user
  * type: discord activity type

  * large_image: url of the large image
  * small_image: url of the small image
  * side_image: url of the side image

  large_image, small_image , side_image are basically the activity images.

<a href="https://api-breeze.herokuapp.com/">API</a>
<a href="https://breeze-stats.netlify.app/">Website</a>

<p align="right">(<a href="#top">back to top</a>)</p>

## Roadmap

- [x] Better client website
- [x] Update svg cards
- [ ] Add more cards for different discord activities
- [x] Endpoints
- [x] Better way for contributors to contribute
- [x] Fix (&) in spotify name bug

<p align="right">(<a href="#top">back to top</a>)</p>

## Issues

One of the biggest issue that i still haven't figured out is with github caching api images.
I'm not sure how to fix this issue.
Would love to get some help from those who know how to fix this issue.

## Contributing

Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.md` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>


<!-- CONTACT -->
## Contact

You can contact me on Discord my username is `Criz#1270`.

<p align="right">(<a href="#top">back to top</a>)</p>


ChangeLog.md
# Change Log

## [1.0.0] - 24.08.2022
Major release
Fixed response time for cards
Fixed bug with large cards
Fixed bug with compact cards

🔨 A lot of code optimization and bug fixes 

Added hex color option for banner 🗼
Made the cards look better and more compact 🔮

Thanks to @codingMASTER398 for the help with the code optimization and the bug fixes

### Todo 
1. Github markdown cache bug fix
2. Better client side code optimization
3. More flexible card design
4. More flexibility with the website to add query parameters
5. Make the cards load faster


server-dcez/lib/api.js
const express = require('express');
const router = express.Router();

const fs = require('fs');
const imageToBase64 = require('image-to-base64');

const { Client } = require('discord.js');
const client = new Client({ intents: 32767 });
require('dotenv').config();

const api = () => {

    router.get('/:id', async (req, res) => {
        // let startTime = Date.now();

        let member
        try {
            member = await client.guilds.cache.get("782646778347388959").members.fetch(req.params.id);
            if (member.user.bot) {
                res.send("User is a bot")
                return
            }
        } catch (e) {

            let html = `
            <html>
            <head>
            <title>User not found</title>
            </head>
            <body>
            <h3>User not found</h3>
            <p>User id: ${req.params.id}</p>
            <p>If the user is not in the DrunkBetch server, please add them to the server and try again</p>
            <a href="https://discord.gg/Ecy6WpEZsD">Join the DrunkBetch server</a>
            <p>If the user is in the server, please try reloading the page</p>
            </body>
            </html>
            `
            res.send(html)
            return;
        }

        let username , discord_avatar, banner, about
        let activity, name, type, details, state
        let large_image, small_image
        let smallimg, raw // for PLaying
        // let spotify_logo, play_along
        let largeText // for Spotify

        let hex
        let temp_small

        let pfp64, banner64
        // let spotify64
        let large64, small64

        try {
            discord_avatar = member.user.displayAvatarURL({ format: 'png', dynamic: true, size: 1024 })
            username = member.user.username + '#' + member.user.discriminator
            banner = req.query.banner
            about = req.query.about || ' '
            if (about.length > 20) {
                about = about.substring(0, 20) + "..."
            }
            
            // spotify_logo = 'https://www.freeiconspng.com/uploads/spotify-icon-0.png'
            // play_along = "https://cdn.discordapp.com/attachments/970974282681307187/987330240609132555/play-along.png"

            temp_small = "https://cdn.discordapp.com/attachments/988140784807202886/991310761991360512/small_breeze.png"
            large_image = req.query.large_image || discord_avatar
            small_image = req.query.small_image || temp_small

            hex = req.query.hex || '4E545B'
        } catch (e) {
            res.send('Uh user error')
            console.log(e)
            return;
        }

        try {
            pfp64 = await imageToBase64(discord_avatar)
            pfp64 = `data:image/png;base64,${pfp64}`

            if (banner) {
                banner64 = await imageToBase64(banner)
                banner64 = `data:image/png;base64,${banner64}`
            } else {
                banner64 = ""
            }

            if(req.query.large_image) {
                large64 = large_image
                large64 = await imageToBase64(large64)
                large64 = `data:image/png;base64,${large64}`
            } else {
                large64 = pfp64
            }
            
            if(req.query.small_image) {
                small64 = small_image
                small64 = await imageToBase64(small64)
                small64 = `data:image/png;base64,${small64}`
            } else {
                small64 = await imageToBase64(temp_small)
                small64 = `data:image/png;base64,${small64}`
            }

        } catch (e) {
            console.log(e)
        }
        
        try {
            if(member.presence.activities[0].id === 'custom' && !member.presence.activities[1]){
                no_activity()
                return;
            } else if (member.presence.activities[0].id === 'custom' || member.presence.activities[0].type === 'CUSTOM') {
                activity = member.presence.activities[1];
            } else {
                try {
                    activity = member.presence.activities[0];
                } catch (error) {
                    no_activity()
                    return;
                }
            }
        } catch (e) {
            let temp;
            function no_activity() {
                try {
                    type = req.query.type || 'Breeze'
                    details = req.query.details || 'Vibing'

                    if (type.length > 23) {
                        type = type.substring(0, 23) + "..."
                    }
                    if (details.length > 23) {
                        details = details.substring(0, 23) + "..."
                    }
                } catch (e) {
                    type = 'Breeze'
                    details = 'Vibing'
                }

                temp = fs.readFileSync('./assets/cards/large/no-activity-new.svg', {encoding: 'utf-8'}).toString()
                temp = temp.replace('[pfp]', pfp64);
                temp = temp.replace('[username]', username);
                temp = temp.replace('[banner]', banner64);
                
                temp = temp.replace('[about]', about);
                temp = temp.replace('[type]', type);
                temp = temp.replace('[details]', details);

                temp = temp.replace('[large-image]', large64);
                // temp = temp.replace('[small-image]', small64);
                temp = temp.replace('[side-image]', small64);

                temp = temp.replace('[hex]', hex);

                res.writeHead(200, {'Content-Type': 'image/svg+xml'})
                res.end(temp)
            }
            no_activity()
            return;
        }
            
        if(!activity.name){
            name = 'No name'
        } else {
            name = activity.name.replace(/&/g, '&amp;');
            if (name.length > 23) {
                name = name.substring(0, 23) + '...';
            }
        }
        
        if(!activity.details){
            details = req.query.details || 'No details'
            if (details.length > 22) {
                details = details.substring(0, 22) + '...';
            }
        } else {
            details = activity.details.replace(/&/g, '&amp;')
            if (details.length > 22) {
                details = details.substring(0, 22) + '...';
            }
        }
        
        if(!activity.state){
            state = req.query.state || 'No state'
            if (state.length > 22) {
                state = state.substring(0, 22) + '...';
            }
        } else {
            state = activity.state.replace(/&/g, '&amp;')
            if (state.length > 22) {
                state = state.substring(0, 22) + '...';
            }
        }
        
        if(!activity.type){
            type = req.query.type || 'No type'
            if (type.length > 19) {
                type = type.substring(0, 19) + '...';
            }
        } else {
            type = activity.type.replace(/&/g, '&amp;')
            if (type.length > 19) {
                type = type.substring(0, 19) + '...';
            }
        }
        
        let temp;
        if (activity.name === 'Spotify') {
            
            let start = activity.timestamps.start
            let end = activity.timestamps.end
            let elapsed = end - start
            let minutes = Math.floor(elapsed / 60000)
            let seconds = Math.floor((elapsed % 60000) / 1000)
            let timeString = `${minutes}:${seconds}` 

            // spotify64 = await imageToBase64(spotify_logo)
            // spotify64 = `data:image/png;base64,${spotify64}`
            // console.log(spotify64)
            // console.log("spotify64 generated in " + (Date.now() - startTime) + "ms")

            if(!activity.assets){
                largeText = req.query.large_text || 'No large text'
            } else if(activity.assets.largeText === null) {
                largeText = req.query.large_text || 'No large text'
            } else {
                largeText = activity.assets.largeText.replace(/&/g, '&amp;');
                if (largeText.length > 23) {
                    largeText = largeText.substring(0, 23) + '...';
                }
            }

            temp = fs.readFileSync('./assets/cards/large/spotify-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[pfp]', pfp64);
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner64);

            temp = temp.replace('[about]', about);
            temp = temp.replace('[details]', details);
            temp = temp.replace('[state]', state);
            temp = temp.replace('[type]', type || 'Vibing');
            temp = temp.replace('[on]', largeText);
            temp = temp.replace('[time]', 'Time -  ' + timeString);
            
            temp = temp.replace('[large-image]', large64);
            temp = temp.replace('[small-image]', small64);

            temp = temp.replace('[hex]', hex);

            temp = temp.replace('[button-text]', "Play on Spotify");
        } else if (activity.name === 'Code' || activity.name === 'Visual Studio Code' ) {

            let time = activity.timestamps.start;
            let elapsed = Date.now() - time;
            let hours = Math.floor(elapsed / 3600000);
            let minutes = Math.floor((elapsed % 3600000) / 60000);
            let seconds = Math.floor((elapsed % 60000) / 1000);
            let timeString = `${hours}:${minutes}:${seconds}`;
            
            const largeimage = activity.assets.largeImage
            let largelink = largeimage.split('raw')[1]
            const rawlarge = 'https://raw' + largelink

            large64 = await imageToBase64(rawlarge)
            large64 = `data:image/png;base64,${large64}`

            const smallimage = activity.assets.smallImage
            let smallink = smallimage.split('raw')[1]
            const rawsmall = 'https://raw' + smallink

            small64 = await imageToBase64(rawsmall)
            small64 = `data:image/png;base64,${small64}`

            temp = fs.readFileSync('./assets/cards/large/vscode-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner64);
            temp = temp.replace('[about]', about);
            temp = temp.replace('[pfp]', pfp64);

            temp = temp.replace('[name]', name || 'Coding');
            temp = temp.replace('[details]', details || 'No details');
            temp = temp.replace('[state]', state || 'No description');
            temp = temp.replace('[type]', type || 'Coding');
            temp = temp.replace('[time]', timeString + ' elapsed' || '0:00 elapsed');

            temp = temp.replace('[large-image]', large64);
            temp = temp.replace('[small-image]', small64);

            temp = temp.replace('[hex]', hex);
            temp = temp.replace('[button-text]', activity.buttons[0] || 'Working on Code');
        } else if (activity.type === 'PLAYING') {

            let time, elapsed, hours, minutes, seconds, timeString
                
            try {
                time = activity.timestamps.start;
                elapsed = Date.now() - time;
                hours = Math.floor(elapsed / 3600000);
                minutes = Math.floor((elapsed % 3600000) / 60000);
                seconds = Math.floor((elapsed % 60000) / 1000);
                timeString = `${hours}:${minutes}:${seconds}`;
            } catch (error) {
                timeString = '0:0:0'
            }

            
            if(!activity.assets){
                raw = req.query.large_image || pfp64
            } else if(activity.assets.smallImage === null) {
                raw = pfp64
            } else if(activity.assets.smallImage.startsWith('mp:external')){
                smallimg = activity.assets.smallImage
                let smalllink = smallimg.split('https/')[1]
                raw = 'https://' + smalllink
                raw = await imageToBase64(raw)
                raw = `data:image/png;base64,${raw}`
            } else {
                raw = req.query.large_image || pfp64
            }

            temp = fs.readFileSync('./assets/cards/large/game-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner64);
            temp = temp.replace('[about]', about);
            temp = temp.replace('[pfp]', pfp64);

            temp = temp.replace('[name]', name || 'Gaming');
            temp = temp.replace('[details]', details);
            temp = temp.replace('[state]', state);
            temp = temp.replace('[type]', type || 'Playing');
            temp = temp.replace('[time]', timeString + ' elapsed' || '0:00 elapsed');

            temp = temp.replace('[large-image]', raw);
            temp = temp.replace('[small-image]', pfp64);
            temp = temp.replace('[hex]', hex);
        }

        res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
        res.setHeader('Pragma', 'no-cache');
        res.setHeader('Expires', '0');
        
        res.writeHead(200, {'Content-Type': 'image/svg+xml'})
        res.end(temp)
        // console.log(`Card generated in ${Date.now() - startTime}ms`);
    })
}

api()
module.exports = router;
client.login(process.env.DISCORD_TOKEN);

server-dcez/lib/base_endpoints/start.js
const express = require('express');
const router = express.Router();

const start = () => {

    router.get('/', async (req, res) => {
        const html = `
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <meta http-equiv="X-UA-Compatible" content="ie=edge">
            <title>Breeze API</title>
            <style>
                body {
                    background-color: #fafafa;
                    display: flex;
                    flex-direction: column;
                    align-items: center;
                    justify-content: center;
                    height: 100vh;
                }
            </style>
        </head>
        <body>
            <h1>Welcome to Breeze API</h1>
            <p>
                The base endpoints are :-
                <ul>
                    <a href="/api">/api</a> <br>
                    <a href="/api/:id">/api/:id</a>
                </ul>
            </p>
        </body>
        </html>
        `;
        res.send(html);
    })
}

start()
module.exports = router;

server-dcez/app.js
const express = require('express');
const app = express();
const fs = require('fs');

const { Client } = require('discord.js');
const Discord = require('discord.js');
const client = new Client({ intents: 32767 });
require('dotenv').config();

const http = require('http')
const { Server } = require('socket.io')
const cors = require('cors');
app.use(cors())

const server = http.createServer(app)

const io = new Server(server, {
    cors: {
        origin: '*',
        methods: ["GET", "POST"],
    }
})

const serverPort = process.env.PORT || 3001;

const start = require('./lib/base_endpoints/start')
app.use('/', start)

const api = require('./lib/api')
app.use('/api/card', api)

const api_xomp = require('./lib/api_xomp')
app.use('/api/compact', api_xomp)

io.on("connection", async (socket) => {
    console.log(`a user connected ${socket.id}`)

    socket.on("user", async function (data) {
        const main_user = data.user
        let member 
        try {
            client.guilds.fetch("782646778347388959")
            member = client.guilds.cache.get('782646778347388959').members.cache.get(main_user) || await client.guilds.cache.get("782646778347388959").members.fetch(main_user);
            if (!member) {
                console.log('member not in server')
                io.emit('not-in-server', {
                    user_id: data.user
                })
                return;
            }
        } catch (error) {
            console.log('member not in server')
            io.emit('not-in-server', {
                user_id: data.user
            })
            return;
        }
        let activity
        try {
            if(member.presence.activities[0].id === 'custom' && !member.presence.activities[1]){
                no_activity()
                return
            } else if (member.presence.activities[0].id === 'custom' || member.presence.activities[0].type === 'CUSTOM') {
                activity = member.presence.activities[1];
            } else {
                try {
                    activity = member.presence.activities[0];
                } catch (error) {
                    no_activity()
                    return;
                }
            }
        } catch (e) {
            no_activity()
            return;
        }
        
        let name, details, state, smallimg, raw, largeText
        let large_image, small_image
        let temp_large, temp_small

        let discord_avatar, username, banner, about
        let play_along, spotify_logo
        
        try {
            discord_avatar = member.user.displayAvatarURL({format: 'png', dynamic: true})
            spotify_logo = 'https://www.freeiconspng.com/uploads/spotify-icon-0.png'
            username = member.user.username + '#' + member.user.discriminator
            banner = data.banner || 'https://media.discordapp.net/attachments/988140784807202886/991308628978061402/blue_boi.png'
            about = data.about || ' '

            if (about.length > 20) {
                about = about.substring(0, 20) + "..."
            }

            play_along = "https://cdn.discordapp.com/attachments/970974282681307187/987330240609132555/play-along.png"
        } catch (e) {
            console.log(e)
            return;
        }
        

        try {
            temp_large = "https://cdn.discordapp.com/attachments/988140784807202886/991310693791965214/large_breeze.png"
            temp_small = "https://cdn.discordapp.com/attachments/988140784807202886/991310761991360512/small_breeze.png"
        } catch (e) {
            console.log(e)
            return;
        }

        if(!activity.name){
            name = ' '
        } else {
            name = activity.name.replace(/&/g, '&amp;');
            if (name.length > 23) {
                name = name.substring(0, 23) + '...';
            }
        }
        
        if(!activity.details){
            details = ' '
        } else {
            details = activity.details.replace(/&/g, '&amp;');
            if (details.length > 23) {
                details = details.substring(0, 23) + '...';
            }
        }

        if(!activity.state){
            state = ' '
        } else {
            state = activity.state.replace(/&/g, '&amp;');
            if (state.length > 23) {
                state = state.substring(0, 23) + '...';
            }
        }

        if(!data.large_image){
            large_image = temp_large 
        } else {
            large_image = data.large_image
        }

        if(!data.small_image){
            small_image = temp_small
        } else {
            small_image = data.small_image
        }

        let temp;
        if (activity.name === 'Spotify') {
            let start = activity.timestamps.start
            let end = activity.timestamps.end
            let elapsed = end - start
            let minutes = Math.floor(elapsed / 60000)
            let seconds = Math.floor((elapsed % 60000) / 1000)
            let timeString = `${minutes}:${seconds}` 

            if(!activity.assets){
                largeText = ' '
            } else if(activity.assets.largeText === null) {
                largeText = ' '
            } else {
                largeText = activity.assets.largeText.replace(/&/g, '&amp;');
                if (largeText.length > 23) {
                    largeText = largeText.substring(0, 23) + '...';
                }
            }

            temp = fs.readFileSync('./assets/cards/large/spotify-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[pfp]', discord_avatar);        
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner);
            
            
            temp = temp.replace('[about]', about);
            temp = temp.replace('[details]', details);
            temp = temp.replace('[state]', state);
            temp = temp.replace('[type]', activity.type);
            temp = temp.replace('[on]', largeText);
            temp = temp.replace('[time]', 'Time -  ' + timeString);
            
            temp = temp.replace('[large-image]', large_image);
            temp = temp.replace('[small-image]', small_image);
            
            temp = temp.replace('[play-along]', play_along);
            temp = temp.replace('[spotify-logo]', spotify_logo);
            temp = temp.replace('[button-text]', "Play on Spotify");
        } else if (activity.name === 'Code' || activity.name === 'Visual Studio Code') {

            let time = activity.timestamps.start;
            let elapsed = Date.now() - time;
            let hours = Math.floor(elapsed / 3600000);
            let minutes = Math.floor((elapsed % 3600000) / 60000);
            let seconds = Math.floor((elapsed % 60000) / 1000);
            let timeString = `${hours}:${minutes}:${seconds}`;
            
            const largeimage = activity.assets.largeImage
            let largelink = largeimage.split('raw')[1]
            const rawlarge = 'https://raw' + largelink

            const smallimage = activity.assets.smallImage
            let smallink = smallimage.split('raw')[1]
            const rawsmall = 'https://raw' + smallink

            temp = fs.readFileSync('./assets/cards/large/vscode-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner);
            temp = temp.replace('[about]', about);
            temp = temp.replace('[pfp]', discord_avatar);

            temp = temp.replace('[name]', activity.name);
            temp = temp.replace('[details]', activity.details);
            temp = temp.replace('[state]', state);
            temp = temp.replace('[type]', activity.type);
            temp = temp.replace('[time]', timeString + ' elapsed');

            temp = temp.replace('[large-image]', rawlarge);
            temp = temp.replace('[small-image]', rawsmall);
            temp = temp.replace('[button-text]', activity.buttons[0] || 'View Repository');
        } else if (activity.type === 'PLAYING') {
            let time, elapsed, hours, minutes, seconds, timeString
            
            try {
                time = activity.timestamps.start;
                elapsed = Date.now() - time;
                hours = Math.floor(elapsed / 3600000);
                minutes = Math.floor((elapsed % 3600000) / 60000);
                seconds = Math.floor((elapsed % 60000) / 1000);
                timeString = `${hours}:${minutes}:${seconds}`;
            } catch (error) {
                timeString = '0:0:0'
            }

            if(!activity.assets){
                raw = discord_avatar
            } else if(activity.assets.smallImage === null) {
                raw = discord_avatar
            } else if(activity.assets.smallImage.startsWith('mp:external')){
                smallimg = activity.assets.smallImage
                let smalllink = smallimg.split('https/')[1]
                raw = 'https://' + smalllink
            } else {
                raw = data.large_image || discord_avatar  
            }
            
            temp = fs.readFileSync('./assets/cards/large/game-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner);
            temp = temp.replace('[about]', about);
            temp = temp.replace('[pfp]', discord_avatar);
    
            temp = temp.replace('[name]', name || 'Gaming');
            temp = temp.replace('[details]', details || '');
            temp = temp.replace('[state]', state || ' ');
            temp = temp.replace('[type]', activity.type || 'PLAYING');
            temp = temp.replace('[large-image]', large_image || raw);
            temp = temp.replace('[small-image]', small_image || raw);
            temp = temp.replace('[time]', timeString || '0:00 elapsed');
        }
        io.emit("message", {
            stuff: activity,
            card: temp,
        })

        function no_activity() {
            let temp

            let discord_avatar, username, banner, about
            let large_image , small_image
            let temp_large, temp_small
            let type , details
                
            try {
                discord_avatar = member.user.displayAvatarURL({format: 'png', dynamic: true})
                username = member.user.username + '#' + member.user.discriminator
                banner = data.banner || 'https://media.discordapp.net/attachments/988140784807202886/991308628978061402/blue_boi.png'
                about = data.about || ' '
                temp_large = "https://cdn.discordapp.com/attachments/988140784807202886/991310693791965214/large_breeze.png"
                temp_small = "https://cdn.discordapp.com/attachments/988140784807202886/991310761991360512/small_breeze.png"

                if (about.length > 20) {
                    about = about.substring(0, 20) + "..."
                }
            } catch (e) {
                console.log(e)
                return;
            }

            if(!data.large_image){
                large_image = temp_large
            } else {
                large_image = data.large_image
            }

            if(!data.small_image){
                small_image = temp_small
            } else {
                small_image = data.small_image
            }

            if(!data.type){
                type = 'Chilling'
            } else {
                type = data.type
            }

            if(!data.details){
                details = 'Breeze'
            } else {
                details = data.details
            }

            temp = fs.readFileSync('./assets/cards/large/no-activity-new.svg', {encoding: 'utf-8'}).toString()
            temp = temp.replace('[pfp]', discord_avatar);
            temp = temp.replace('[username]', username);
            temp = temp.replace('[banner]', banner);
            temp = temp.replace('[about]', about);

            temp = temp.replace('[type]', type);
            temp = temp.replace('[details]', details);

            temp = temp.replace('[large-image]', large_image || discord_avatar)
            temp = temp.replace('[small-image]', small_image || discord_avatar)
            temp = temp.replace('[side-image]', discord_avatar)

            // console.log(`${member.user.username} has no activity`)

            io.emit('no-activity', {
                user_id: data.user,
                card: temp
            })
            return
        }
    })
})

io.on("disconnect", (socket) => {
    console.log(`a user disconnected ${socket.id}`)
})

process.on('unhandledRejection', async (reason, p, origin) => {
    const embed = new Discord.MessageEmbed()
        .setTitle('Error Occured')
        .setColor('RANDOM')
        .setDescription('```js\n' + reason.stack + '```');
    client.channels.cache.get('988140784807202886').send({ embeds: [embed] })
});

process.on('uncaughtExceptionMonitor', async (err, origin) => {
    const embed = new Discord.MessageEmbed()
        .setTitle('Error Occured')
        .setColor('RANDOM')
        .setDescription('```js\n' + err.stack + '```');
    client.channels.cache.get('988140784807202886').send({ embeds: [embed] })
});

server.listen(process.env.PORT || serverPort , () => console.log(`Listening on port ${process.env.PORT || serverPort}`))
// console.log(`http://localhost:3001/api/card/784141856426033233`)
// console.log(`http://localhost:3001/api/compact/784141856426033233`)
client.login(process.env.DISCORD_TOKEN);